### PR TITLE
fix(filetype): make .tex filetype detection match Vim

### LINKED ISSUE #21594: Error opening LaTeX files with precompiled headers.
### Neovim version (nvim -v)

0.8.2

### Vim (not Nvim) behaves the same?

no (problem with detect.lua)

### Operating system/version

Arch 6.1.1-arch1-1

### Terminal name/version

Konsole 22.12.0

### $TERM environment variable

xterm-256color

### Installation

pacman (system package manager)

### How to reproduce the issue

Open any latex file that uses pre-compiled headers. The first line of this file will be

    %& preamble

### Expected behavior

File should be opened, and detected as Latex

### Actual behavior

The filetype is not detected, and you get a few error messages from the runtime file `filetype/detect.lua`. (I fixed the problem and am pasting the fix below, so I can't reproduce the error messages, sorry.)

The fix is to edit `filetype/detect.lua` line 1245 (the function `M.tex`), and change the first few lines to:

    function M.tex(path, bufnr)
      local found, idx, format = getlines(bufnr, 1):find('^%%&%s*(%a+)')
      if found then
        format = format:lower():gsub('pdf', '', 1)
        if format == 'plaintex' then
          return 'plaintex'
        else
          return 'tex'
        end
      elseif ... (everything else is unchanged)

Note, in addition to fixing the runtime error messages, it changed the detection logic. `%&plaintex` files get treated as plaintex, and everything else gets treated as LaTeX. This includes `%&tex`, and those with pre-compiled  headers.

### PR DESCRIPTION
**Pull request by gi1242**
Fix #21594

Synced detection logic to match Vim9 for files with a pre-compilied header.

### CODE REVIEW

File: runtime/lua/vim/filetype/detect.lua [hunk 1]
```diff
@@ -1256,17 +1256,16 @@ end
...
 function M.tex(path, bufnr)
-  local format = getlines(bufnr, 1):find('^%%&%s*(%a+)')
-  if format then
-    format = format:lower():gsub('pdf', '', 1)
-    if format == 'tex' then
-      return 'tex'
-    elseif format == 'plaintex' then
-      return 'plaintex'
-    end
+  local matched, idx, mstr = getlines(bufnr, 1):find('^%%&%s*(%a+)')
```
Comment (smjonas) [inline runtime/lua/vim/filetype/detect.lua:1259]: Thanks for the PR! I would prefer to only change this line to `local _, _, format = getlines(bufnr, 1):find('^%%&%s*(%a+)')` though. The other changes such as directly returning the filetype were made deliberately to simplify the logic compared to Vim's implementation.

**Review by smjonas — CHANGES_REQUESTED** [review]

**Comment by zeertzjq** [comment]
test failure

**Comment by clason** [comment]
Specifically, failure of Vim's filetype detection test, which means that the detection logic changed.

File: runtime/lua/vim/filetype/detect.lua (hunk 1 above)
Comment (gi1242) [reply runtime/lua/vim/filetype/detect.lua:1259]: The other changes you had unfortunately ended up changing the logic! If format didn't match in the first few lines, the fall back to `latex` is done by Vim, but not by the neovim you had before.

Just try whatever code you end up using on a file that has `%&preamble` in the first file. It should be detected as LaTeX (Vim does it).

Comment (clason) [reply runtime/lua/vim/filetype/detect.lua:1259]: That's fine, but the tests (for more typical situations) -- which are taken verbatim from vim! -- must pass as well.

**Comment by gi1242** [comment]
Sorry, I'm not familiar with the build process, so can't see what test failed. Can I suggest adding a file `precompiled.tex` with contents:

    %& preamble
    \begin{document}
      Stuff
    \end{document}
    
Vim opens it as LaTeX (as it should), and neovim did not. This patch fixes it.

**Comment by zeertzjq** [comment]
The test failure is for `tex/context/.*/.*.tex`

**Comment by gi1242** [comment]
I got your build system running on my machine and tried making tests. Can I get some help. I tried various combinations of `TEST_FILTER="..." make functionaltest` and couldn't narrow it down to the test that failed. Just running `make functionaltest` says

    CMake Error at /home/gautam/tmp/neovim/cmake/RunTests.cmake:107 (message):
      functional tests failed with error: 1


    FAILED: CMakeFiles/functionaltest /home/gautam/tmp/neovim/build/CMakeFiles/functionaltest
    
I couldn't find log files, or instructions on how to find the test that failed...

Thanks

**Comment by zeertzjq** [comment]
You don't need to run functionaltests. You need to run `make oldtests TEST_FILE=test_filetype.vim`

**Comment by gi1242** [comment]
Thanks. I had a typo (wrote `contex` instead of `context`. My bad. Tests pass now (on my system)

**Review by zeertzjq — DISMISSED** [review]

File: runtime/lua/vim/filetype/detect.lua [hunk 2]
```diff
@@ -1279,22 +1278,21 @@ function M.tex(path, bufnr)
...
-    elseif vim.g.tex_flavor == 'context' then
-      return 'context'
-    else
-      -- Probably LaTeX
-      return 'tex'
-    end
+  end -- if matched
+
+  -- Translation from formats to file types.  TODO:  add AMSTeX, RevTex, others?
+  if format == 'plaintex' or format == 'context' then
```
Comment (zeertzjq) [inline runtime/lua/vim/filetype/detect.lua:1291]: This changes the behavior of `g:tex_flavor = 'plain'`. With `g:tex_flavor = 'plain'` filetype should be set `plaintex`, but with this PR it is set to `tex`.

**Comment by smjonas** [comment]
I have created a PR on the Vim repo that adds more test cases: https://github.com/vim/vim/pull/11765. This should contain all things mentioned in this thread.

File: runtime/lua/vim/filetype/detect.lua (hunk 2 above)
Comment (gi1242) [reply runtime/lua/vim/filetype/detect.lua]: You're right, my apologies. I just pushed a fix.

**Comment by clason** [comment]
@gi1242 could you cherry-pick https://github.com/neovim/neovim/pull/21601/commits/cc70649f6882fc6b12badcaeffd5cc93eeafe6eb (and squash your other commits?)

File: runtime/lua/vim/filetype/detect.lua (hunk 1 above)
Comment (gi1242) [reply runtime/lua/vim/filetype/detect.lua]: fixed now

**Comment by gi1242** [comment]
Done... (I think. I had to force push, and hopefully it updates here. If not it's f3e23ee2e3 in gi1242/neovim). Thanks for helping me out with your build system

**Comment by clason** [comment]
That squash worked -- but was too eager, the vim-patch commit should have been left separate!

**Comment by gi1242** [comment]
10th time a charm?

**Comment by clason** [comment]
Sorry, still not quite: you need to cherry-pick the commit from _my_ PR, otherwise it doesn't count as a ported Vim patch.

**Comment by gi1242** [comment]
Ooof. Fixed...

**Comment by clason** [comment]
Test failure:
```
From test_filetype.vim:
[4818](https://github.com/neovim/neovim/actions/runs/3812943263/jobs/6486418411#step:14:4819)
Found errors in Test_tex_file():
[4819](https://github.com/neovim/neovim/actions/runs/3812943263/jobs/6486418411#step:14:4820)
command line..script /home/runner/work/neovim/neovim/src/nvim/testdir/runtest.vim[437]..function RunTheTest[44]..Test_tex_file line 10: Expected 'tex' but got 'plaintex'
[4820](https://github.com/neovim/neovim/actions/runs/3812943263/jobs/6486418411#step:14:4821)
command line..script /home/runner/work/neovim/neovim/src/nvim/testdir/runtest.vim[437]..function RunTheTest[44]..Test_tex_file line 32: Expected 'context' but got 'tex'
[4821](https://github.com/neovim/neovim/actions/runs/3812943263/jobs/6486418411#step:14:4822)
/home/runner/work/neovim/neovim//build/bin/nvim -u NONE -U NONE -i NONE --noplugin --headless -S summarize.vim messages
```

**Comment by smjonas** [comment]
Do we really need a separate `format` variable as of now? Imo this adds unnecessary complexity. It might make sense to readd it if more formats get added in the future. To map `vim.g.tex_format` to the correct filetype, we can use a table. 

Minor suggestion: use the `format` variable directly instead of `mstr` or at least rename it to make the meaning clearer?

**Comment by gi1242** [comment]
Sorry to bug you guys about this so often. I'm still having trouble with your test system. Looks like the test on line 10 failed. That test creates a file called `Xfile.tex` with contents

    \newcommand{\test}{some text}
    
and then requires the filetype is `tex`. When the test runs it appears to give `plaintex` as the filetype, and I couldn't figure out why. So I created the above file, and opened it with

    install/bin/nvim -R -u NONE --cmd 'filetype on' ~/tmp/Xfile.tex -c 'echo &filetype'
    
and I got the filetype to be `tex` ...

If you know what's going on, or have other suggestions, please let me konw. I seem to have a better idea of what you expect now, so hopefully I can get a pull request in the right format once I figure out how to debug the test... 😄

**Comment by gi1242** [comment]
Never mind. I see the problem. `g:tex_flavor` seems to be set when I run `nvim -u NONE`, but not in the test. I'll push a fix soon (with working tests)

**Comment by gi1242** [comment]
Ok. All tests should pass now, and I renamed `mstr` to `format` as you requested. (I can't seem to avoid it entirely.)

Just FYI, the context detection previously there was also implemented wrong. The Vim used the 'p' search flag which made the return value the matching subexpression. This doesn't seem possible with the lua `matchstring` function, and as far as I could tell it would never match. Not to mention the starting line offset was off by one.

I think I have everything fixed now.

It also WorksForMe™, so 😁

File: runtime/lua/vim/filetype/detect.lua [hunk 3]
```diff
@@ -1256,45 +1256,41 @@ end
...
     for i, l in ipairs(getlines(bufnr, 1, 1000)) do
       -- Find first non-comment line
       if not l:find('^%s*%%%S') then
         -- Check the next thousand lines for a LaTeX or ConTeXt keyword.
-        for _, line in ipairs(getlines(bufnr, i + 1, i + 1000)) do
-          local lpat_match, cpat_match =
-            matchregex(line, [[\c^\s*\\\%(]] .. lpat .. [[\)\|^\s*\\\(]] .. cpat .. [[\)]])
-          if lpat_match then
+        for _, line in ipairs(getlines(bufnr, i, i + 1000)) do
+          if matchregex(line, [[\c^\s*\\\%(]] .. lpat .. [[\)]] ) then
```
Comment (zeertzjq) [inline runtime/lua/vim/filetype/detect.lua:1276]:
```suggestion
          if matchregex(line, [[\c^\s*\\\%(]] .. lpat .. [[\)]]) then
```

File: runtime/lua/vim/filetype/detect.lua [hunk 4]
```diff
@@ -1256,45 +1256,41 @@ end
...
             return 'tex'
-          elseif cpat_match then
+          elseif matchregex(line, [[\c^\s*\\\%(]] .. cpat .. [[\)]] ) then
```
Comment (zeertzjq) [inline runtime/lua/vim/filetype/detect.lua:1278]:
```suggestion
          elseif matchregex(line, [[\c^\s*\\\%(]] .. cpat .. [[\)]]) then
```

**Review by zeertzjq — COMMENTED** [review]
lint failure

**Comment by gi1242** [comment]
Lint failure again... Looks never ending. Pushed hopefully the last fix.

**Review by smjonas — APPROVED** [review]

File: runtime/lua/vim/filetype/detect.lua [hunk 5]
```diff
@@ -1256,17 +1256,15 @@ end
...
-    if format == 'tex' then
-      return 'tex'
-    elseif format == 'plaintex' then
-      return 'plaintex'
-    end
   elseif path:lower():find('tex/context/.*/.*%.tex') then
     return 'context'
   else
+    -- Default value, may be changed later:
+    format = vim.g.tex_flavor and vim.g.tex_flavor or 'plaintex'
```
Comment (smjonas) [inline runtime/lua/vim/filetype/detect.lua:1266]: Can be simplified to `format = vim.g.tex_flavor or 'plainex'`

Comment (gi1242) [reply runtime/lua/vim/filetype/detect.lua]: Indeed, done (force pushed)

**Review by zeertzjq — APPROVED** [review]

**Pull request merged**